src/nanoemoji/write_font2svg.py

```diff
@@ -0,0 +1,94 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generate SVG files from the SVG font table."""
+from absl import app
+from absl import flags
+from absl import logging
+from fontTools import ttLib
+from lxml import etree
+from nanoemoji import codepoints
+from nanoemoji.color_glyph import map_viewbox_to_otsvg_space
+from nanoemoji import util
+import os
+import shutil
+import subprocess
+import sys
+from pathlib import Path
+from picosvg.svg import SVG
+from picosvg.svg_meta import strip_ns
+
+
+FLAGS = flags.FLAGS
+
+
+flags.DEFINE_string(
+    "output_dir", None, "Output dir. Files written to <glyph name>.svg."
+)
+flags.DEFINE_string(
+    "log_level",
+    "INFO",
+    "The threshold for what messages will be logged. One of DEBUG, INFO, WARN, "
+    "ERROR, or FATAL.",
+)
+
+
+def main(argv):
+    logging.set_verbosity(FLAGS.log_level)
+
+    font_file = util.only(lambda a: a.endswith(".ttf"), argv)
+    out_dir = Path(FLAGS.output_dir)
+    assert out_dir.is_dir(), f"{FLAGS.output_dir} is not a directory"
+
+    font = ttLib.TTFont(font_file)
+    assert "SVG " in font, f"No SVG table in {font_file}"
+    upem = font["head"].unitsPerEm
+    ascender = font["OS/2"].sTypoAscender
+    descender = font["OS/2"].sTypoDescender
+    metrics = font["hmtx"].metrics
+    logging.debug("Writing svgs from %s with to %s", font_file, out_dir)
+    logging.debug("upem %d ascender %d descender %d", upem, ascender, descender)
+
+    # We want a subsequent nanoemoji scale to be 1:1
+    # So use the font height (global) and width (per glyph) as the svg viewbox
+    height = ascender - descender
+
+    for raw_svg, min_gid, max_gid in font["SVG "].docList:
+        svg = SVG.fromstring(raw_svg)
+        svg_defs = etree.Element("defs")
+        svg_g = etree.Element("g")
+        svg_g.attrib["transform"] = f"translate(0, {ascender})"
+
+        for el in svg.svg_root:
+            if strip_ns(el.tag) == "defs":
+                svg_defs.append(el)
+            else:
+                svg_g.append(el)
+
+        svg.svg_root.append(svg_defs)
+        svg.svg_root.append(svg_g)
+
+        for gid in range(min_gid, max_gid + 1):
+            glyph_name = font.getGlyphName(gid)
+            width, _ = metrics[glyph_name]
+            svg.svg_root.attrib["viewBox"] = f"0 0 {width} {height}"
+            dest_file = out_dir / f"{glyph_name}.svg"
+            with open(dest_file, "w") as f:
+                f.write(svg.tostring(pretty_print=True))
```

Comment:
so, you're writing the _same_ SVG object to multiple files, with different filenames and possibly different width. That's not correct - it may look ok for simple fonts like the one you tested where there's one glyph per svg document; but for those like the `picosvg` ones we generate, where there's shape reuse and multiple glyphs bunlded in the same doc, it will produce incorrect results.
What you want to do, in the general case is the following:

1) For each glyph name you need to create a separate SVG object
2) move all the elements (be it `<defs>` or anything) that are _not_ the `<g>` for the given glyph that you are considering, into a `<defs>` element at the top; this way they won't be displayed, but can be still referenced 
3) only keep the current `<g>` element at the top-level following (outside of) the global `<defs>`; this way only that glyph will be displayed
4) optionally, drop all stuff in `<defs>` which is not referenced (reachable via xlink:href or gradient/clip-path refs) from the current `<g>` glyph we are drawing (similarly to how the fonttools subsetter does).

Reply:
Support for multi-glyph svgs coming soon :) Does the subsetter have reusable APIs for this at all? - seems a bit of a shame to reimplement discarding glyphs and treeshaking.

Reply:
I added deletion of elements for all ids but the one we are keeping. The result gets sent through picosvg which should strip unused elements.